Migrate GetCertAuthority to gRPC

rosstimothy opened the pull request:
The HTTP client created by `auth.NewClient` has an idle timeout of
30s and each instance periodically gets cert authorities in
`TeleportProcess.syncRotationStateCycle`. If there has been no
 activity when the ticker expires then the request will result in a
brand new connection to Auth to be established. This can cause a
spike in open file descriptors on the Proxy when there are a large
number of instances connected via a tunnel.

To prevent periodic dialing when retrieving cert authorities the
endpoint has been migrated to the new Trust gRPC service. For now
only the get endpoint has been migrated but the rest of the CRUD
operations for cert authorities should be migrated from the HTTP
api to the gRPC service as well.

fspmarshall commented on api/client/client.go:
```diff
@@ -611,6 +612,31 @@ func (c *Client) LoginRuleClient() loginrulepb.LoginRuleServiceClient {
+	return trustpb.NewTrustServiceClient(c.conn)
+}
+
+// closedClientConnection is a [grpc.ClientConnInterface] implementation
+// that returns a connection error when a unary RPC is invoked or a new
+// stream is attempted to be created.
+type closedClientConnection struct{}
+
+// Invoke implements [grpc.ClientConnInterface]
+func (g closedClientConnection) Invoke(ctx context.Context, method string, args interface{}, reply interface{}, opts ...grpc.CallOption) error {
+	return trace.ConnectionProblem(nil, "connection was closed when attempting to call %s", method)
+}
+
+// NewStream implements [grpc.ClientConnInterface]
+func (g closedClientConnection) NewStream(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	return nil, trace.ConnectionProblem(nil, "connection was closed when attempting to call %s", method)
+}
```
nit: IIUC the purpose of this type is to be a placeholder for building sub-clients from an already closed root client.  I'd clarify that purpose on the comments, and update the error message to indicate that the client itself was explicitly closed.

"connection was closed while [...]" feels like it could be a network error, or possibly something the remote server did, but something like "cannot call <method> after client has already been closed" would make me immediately check where I call `Client.Close`.  Easier to debug.

rosstimothy committed:
Migrate GetCertAuthority to gRPC

The HTTP client created by `auth.NewClient` has an idle timeout of
30s and each instance periodically gets cert authorities in
`TeleportProcess.syncRotationStateCycle`. If there has been no
activity when the ticker expires then the request will result in a
brand new connection to Auth to be established. This can cause a
spikes in open file descriptors on the Proxy when there are a large
number of instances connected via a tunnel.

To prevent periodic dialing when retrieving cert authorities the
endpoint has been migrated to the new Trust gRPC service. For now
only the get endpoint has been migrated but the rest of the CRUD
operations for cert authorities should be migrated from the HTTP
api to the gRPC service as well.
